File: shared/utils/broadcast_schedule.py
"""Broadcast schedule helpers — stored slots are Japan Standard Time (JST)."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone

JST = timezone(timedelta(hours=9))
_WEEKDAY_SHORT = ("Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun")


@dataclass(frozen=True)
class BroadcastSlot:
    """Weekly airing slot stored as JST wall-clock time."""

    weekday: int  # 0=Monday .. 6=Sunday
    hour: int
    minute: int


def parse_broadcast(value: str | None) -> BroadcastSlot | None:
    """Parse a ``weekday-hour-minute`` broadcast string."""
    if not value:
        return None
    parts = str(value).strip().split("-")
    if len(parts) != 3:
        return None
    try:
        weekday, hour, minute = (int(part) for part in parts)
    except ValueError:
        return None
    if not 0 <= weekday <= 6:
        return None
    if not 0 <= hour <= 23:
        return None
    if not 0 <= minute <= 59:
        return None
    return BroadcastSlot(weekday=weekday, hour=hour, minute=minute)


def _resolve_tz(tz: timezone | None) -> timezone:
    if tz is not None:
        return tz
    local = datetime.now().astimezone().tzinfo
    if isinstance(local, timezone):
        return local
    return timezone.utc


def convert_jst_slot_to_local(
    slot: BroadcastSlot,
    tz: timezone | None = None,
    now: datetime | None = None,
) -> BroadcastSlot:
    """Convert a recurring JST slot to local wall-clock weekday/time."""
    tzinfo = _resolve_tz(tz)
    local_next = next_episode_datetime(slot, now=now, tz=tzinfo)
    return BroadcastSlot(
        weekday=local_next.weekday(),
        hour=local_next.hour,
        minute=local_next.minute,
    )


def format_slot_short(slot: BroadcastSlot) -> str:
    return f"{_WEEKDAY_SHORT[slot.weekday]} {slot.hour:02d}:{slot.minute:02d}"


def format_broadcast_jst(slot: BroadcastSlot) -> str:
    return f"{format_slot_short(slot)} JST"


def format_broadcast_display(
    slot: BroadcastSlot,
    tz: timezone | None = None,
    *,
    include_jst: bool = True,
    now: datetime | None = None,
) -> str:
    """Render a broadcast slot in local time, with optional JST annotation."""
    local = convert_jst_slot_to_local(slot, tz, now=now)
    local_text = format_slot_short(local)
    if not include_jst:
        return local_text
    jst_text = format_slot_short(slot)
    if local_text == jst_text:
        return local_text
    return f"{local_text} ({jst_text} JST)"


def next_episode_datetime(
    slot: BroadcastSlot,
    now: datetime | None = None,
    tz: timezone | None = None,
) -> datetime:
    """Return the next airing datetime in the requested timezone."""
    tzinfo = _resolve_tz(tz)
    now_jst = (now or datetime.now(timezone.utc)).astimezone(JST)

    days_ahead = (slot.weekday - now_jst.weekday()) % 7
    candidate = now_jst.replace(
        hour=slot.hour, minute=slot.minute, second=0, microsecond=0
    ) + timedelta(days=days_ahead)
    if candidate <= now_jst:
        candidate += timedelta(days=7)
    return candidate.astimezone(tzinfo)


def latest_episode_label(
    slot: BroadcastSlot,
    now: datetime | None = None,
    tz: timezone | None = None,
) -> str:
    """Return a relative label for the most recent weekly episode."""
    tzinfo = _resolve_tz(tz)
    now_local = (now or datetime.now(timezone.utc)).astimezone(tzinfo)
    last_local = next_episode_datetime(slot, now, tzinfo) - timedelta(days=7)
    days_since = (now_local.date() - last_local.date()).days

    if days_since == 0:
        return "Today"
    if days_since == 1:
        return "Yesterday"
    if days_since > 1:
        return f"{days_since} days ago"
    return "uhh?"


def utc_timestamp_to_jst_slot(timestamp: int) -> BroadcastSlot:
    """Convert a UTC unix timestamp into a recurring JST slot."""
    dt_jst = datetime.fromtimestamp(int(timestamp), timezone.utc).astimezone(JST)
    return BroadcastSlot(
        weekday=dt_jst.weekday(),
        hour=dt_jst.hour,
        minute=dt_jst.minute,
    )


def broadcast_slot_to_string(slot: BroadcastSlot) -> str:
    return f"{slot.weekday}-{slot.hour}-{slot.minute}"

File: next-web/lib/broadcast-schedule.ts
/** Broadcast schedule helpers — stored slots are Japan Standard Time (JST). */

export type BroadcastSlot = {
  weekday: number; // 0=Monday .. 6=Sunday
  hour: number;
  minute: number;
};

const WEEKDAY_SHORT = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"] as const;

type ZonedParts = {
  year: number;
  month: number;
  day: number;
  hour: number;
  minute: number;
  second: number;
  weekday: number;
};

export function parseBroadcast(value?: string | null): BroadcastSlot | null {
  if (!value) return null;
  const parts = value.trim().split("-");
  if (parts.length !== 3) return null;
  const weekday = Number.parseInt(parts[0] ?? "", 10);
  const hour = Number.parseInt(parts[1] ?? "", 10);
  const minute = Number.parseInt(parts[2] ?? "", 10);
  if (!Number.isFinite(weekday) || weekday < 0 || weekday > 6) return null;
  if (!Number.isFinite(hour) || hour < 0 || hour > 23) return null;
  if (!Number.isFinite(minute) || minute < 0 || minute > 59) return null;
  return { weekday, hour, minute };
}

function zonedParts(date: Date, timeZone: string): ZonedParts {
  const formatter = new Intl.DateTimeFormat("en-US", {
    timeZone,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    weekday: "short",
    hourCycle: "h23",
  });
  const parts = formatter.formatToParts(date);
  const read = (type: Intl.DateTimeFormatPartTypes) =>
    Number.parseInt(parts.find((part) => part.type === type)?.value ?? "0", 10);
  const weekdayLabel = parts.find((part) => part.type === "weekday")?.value ?? "Mon";
  const weekday = WEEKDAY_SHORT.indexOf(weekdayLabel as (typeof WEEKDAY_SHORT)[number]);
  return {
    year: read("year"),
    month: read("month"),
    day: read("day"),
    hour: read("hour"),
    minute: read("minute"),
    second: read("second"),
    weekday: weekday >= 0 ? weekday : 0,
  };
}

function localTimeInZoneToUtc(
  year: number,
  month: number,
  day: number,
  hour: number,
  minute: number,
  second: number,
  timeZone: string,
): number {
  let guess = Date.UTC(year, month - 1, day, hour, minute, second);
  for (let adjust = -14; adjust <= 14; adjust++) {
    const ms = guess + adjust * 3_600_000;
    const parts = zonedParts(new Date(ms), timeZone);
    if (
      parts.year === year &&
      parts.month === month &&
      parts.day === day &&
      parts.hour === hour &&
      parts.minute === minute &&
      parts.second === second
    ) {
      return ms;
    }
  }
  return guess;
}

export function convertJstSlotToLocal(
  slot: BroadcastSlot,
  timeZone: string,
  now: Date = new Date(),
): BroadcastSlot {
  const next = nextEpisodeDate(slot, now, timeZone);
  const parts = zonedParts(next, timeZone);
  return {
    weekday: parts.weekday,
    hour: parts.hour,
    minute: parts.minute,
  };
}

export function formatSlotShort(slot: BroadcastSlot): string {
  const weekday = WEEKDAY_SHORT[slot.weekday] ?? "???";
  return `${weekday} ${String(slot.hour).padStart(2, "0")}:${String(slot.minute).padStart(2, "0")}`;
}

export function formatBroadcastJst(slot: BroadcastSlot): string {
  return `${formatSlotShort(slot)} JST`;
}

export function formatBroadcastDisplay(
  slot: BroadcastSlot,
  timeZone: string,
  options: { includeJst?: boolean; now?: Date } = {},
): string {
  const includeJst = options.includeJst ?? true;
  const now = options.now ?? new Date();
  const local = convertJstSlotToLocal(slot, timeZone, now);
  const localText = formatSlotShort(local);
  if (!includeJst) return localText;
  const jstText = formatSlotShort(slot);
  if (localText === jstText) return localText;
  return `${localText} (${jstText} JST)`;
}

export function nextEpisodeDate(
  slot: BroadcastSlot,
  now: Date = new Date(),
  timeZone: string,
): Date {
  const nowMs = now.getTime();
  for (let offset = 0; offset <= 14; offset++) {
    const probe = new Date(nowMs + offset * 86_400_000);
    const jst = zonedParts(probe, "Asia/Tokyo");
    if (jst.weekday !== slot.weekday) continue;

    const candidateMs = localTimeInZoneToUtc(
      jst.year,
      jst.month,
      jst.day,
      slot.hour,
      slot.minute,
      0,
      "Asia/Tokyo",
    );
    if (candidateMs > nowMs) {
      return new Date(candidateMs);
    }
  }

  const fallbackMs = localTimeInZoneToUtc(
    zonedParts(now, "Asia/Tokyo").year,
    zonedParts(now, "Asia/Tokyo").month,
    zonedParts(now, "Asia/Tokyo").day,
    slot.hour,
    slot.minute,
    0,
    "Asia/Tokyo",
  );
  return new Date(fallbackMs + 7 * 86_400_000);
}

export function formatNextEpisodeLine(
  slot: BroadcastSlot,
  now: Date = new Date(),
  timeZone: string,
): string {
  const next = nextEpisodeDate(slot, now, timeZone);
  const formatter = new Intl.DateTimeFormat("en-US", {
    timeZone,
    weekday: "short",
    day: "numeric",
    hour: "2-digit",
    minute: "2-digit",
    hourCycle: "h23",
  });
  return `Next episode on ${formatter.format(next)}`;
}

function calendarDayKey(date: Date, timeZone: string): string {
  return new Intl.DateTimeFormat("en-CA", {
    timeZone,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).format(date);
}

export function latestEpisodeLabel(
  slot: BroadcastSlot,
  now: Date = new Date(),
  timeZone: string,
): string {
  const next = nextEpisodeDate(slot, now, timeZone);
  const last = new Date(next.getTime() - 7 * 86_400_000);
  const nowDay = calendarDayKey(now, timeZone);
  const lastDay = calendarDayKey(last, timeZone);

  if (nowDay === lastDay) return "Today";

  const nowUtc = Date.parse(`${nowDay}T12:00:00Z`);
  const lastUtc = Date.parse(`${lastDay}T12:00:00Z`);
  const daysSince = Math.round((nowUtc - lastUtc) / 86_400_000);

  if (daysSince === 1) return "Yesterday";
  if (daysSince > 1) return `${daysSince} days ago`;
  return "uhh?";
}

export function buildBroadcastAiringLines(
  broadcast: string | null | undefined,
  timeZone: string,
  now: Date = new Date(),
): string[] {
  const slot = parseBroadcast(broadcast);
  if (!slot) return [];
  return [
    formatNextEpisodeLine(slot, now, timeZone),
    `Latest episode: ${latestEpisodeLabel(slot, now, timeZone)}`,
  ];
}

export function mergeAiringLines(
  serverLines: string[],
  broadcast: string | null | undefined,
  timeZone: string,
  now: Date = new Date(),
): string[] {
  const base = serverLines.filter(
    (line) =>
      !line.startsWith("Next episode on") && !line.startsWith("Latest episode:"),
  );
  if (!broadcast) return base;
  return [...base, ...buildBroadcastAiringLines(broadcast, timeZone, now)];
}
